core/integration/remotecache_test.go

```diff
@@ -3,252 +3,193 @@ package core
-		*/
-		clientA := getClient()
-		clientB := getClient()
-		outputA := pipelineOutput(clientA)
-		require.NoError(t, clientA.Close())
-		outputB := pipelineOutput(clientB)
-		require.Equal(t, outputA, outputB)
-	})
+	return devEngine, endpoint, err
 }
 
-func runS3InDocker(ctx context.Context, t *testing.T, bucket string) string {
-	t.Helper()
-	name := "dagger-test-remote-cache-s3-" + identity.NewID()
-	cmd := exec.Command("docker", "run", "--rm", "--name", name, "minio/minio", "server", "/data")
-	t.Cleanup(func() {
-		stopDockerRun(cmd, name)
-	})
-	cmd.Stdout = os.Stdout
-	cmd.Stderr = os.Stderr
-	err := cmd.Start()
+func TestRemoteCacheRegistry(t *testing.T) {
+	c, ctx := connect(t)
+	defer c.Close()
+
+	registry := c.Pipeline("registry").Container().From("registry:2").
+		WithExposedPort(5000, dagger.ContainerWithExposedPortOpts{Protocol: dagger.Tcp}).
+		WithExec(nil)
+
+	devEngine, endpoint, err := getDevEngine(ctx, c, registry, "registry", "type=registry,ref=registry:5000/test-cache,mode=max", 0)
 	require.NoError(t, err)
-	// wait for the s3 to be ready
-	for i := 0; i < 100; i++ {
-		cmd := exec.CommandContext(ctx, "docker", "exec", name, "sh", "-c", "curl -s -o /dev/null -w '%{http_code}' http://localhost:9000/minio/health/live")
-		out, err := cmd.CombinedOutput()
-		if string(out) == "200" && err == nil {
-			break
-		}
-		if i == 99 {
-			t.Fatalf("minio s3 not ready: %v: %s", err, out)
-		}
-		time.Sleep(100 * time.Millisecond)
-	}
 
-	// create the bucket
-	cmd = exec.Command("docker", "run", "--rm", "--network", "container:"+name, "--entrypoint", "sh", "minio/mc", "-c", "mc config host add minio http://localhost:9000 minioadmin minioadmin && mc mb minio/"+bucket) // #nosec G204
-	cmd.Stdout = os.Stdout
-	cmd.Stderr = os.Stderr
-	err = cmd.Run()
+	// This loads the dagger-cli binary from the host into the container, that was set up by
+	// internal/mage/engine.go:test. This is used to communicate with the dev engine.
+	daggerCli := c.Host().Directory("/dagger-dev/", dagger.HostDirectoryOpts{Include: []string{"dagger"}}).File("dagger")
+
+	cliBinPath := "/.dagger-cli"
+
+	outputA, err := c.Container().From("alpine:3.17").
+		WithServiceBinding("dev-engine", devEngine).
+		WithMountedFile(cliBinPath, daggerCli).
+		WithEnvVariable("_EXPERIMENTAL_DAGGER_CLI_BIN", cliBinPath).
+		WithEnvVariable("_EXPERIMENTAL_DAGGER_RUNNER_HOST", endpoint).
+		WithNewFile("/.dagger-query.txt", dagger.ContainerWithNewFileOpts{
+			Contents: `{ 
+				container { 
+					from(address: "alpine:3.17") { 
+						withExec(args: ["sh", "-c", "head -c 128 /dev/random | sha256sum"]) { 
+							stdout 
+						} 
+					} 
+				} 
+			}`}).
+		WithExec([]string{
+			"sh", "-c", cliBinPath + ` query --doc .dagger-query.txt`,
+		}).Stdout(ctx)
 	require.NoError(t, err)
+	shaA := strings.TrimSpace(gjson.Get(outputA, "container.from.exec.stdout").String())
```

Comment:
Here and on the line for shaB, I think this needs to be:
```suggestion
	shaA := strings.TrimSpace(gjson.Get(outputA, "container.from.withExec.stdout").String())
```

I think right now both SHAs end up being `""`

Reply:
Good catch. I'll add an assert that they aren't empty, too.